src/IdentityManager/DataService/IdentityServices/CustomPersistedGrantStore.cs

```diff
@@ -0,0 +1,154 @@
+﻿using IdentityManager.DataService.Mappers;
+using IdentityServer4.Extensions;
+using IdentityServer4.Models;
+using IdentityServer4.Stores;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Travely.IdentityManager.Repository.Abstractions;
+
+namespace IdentityManager.DataService.IdentityServices
+{
+    public class CustomPersistedGrantStore : IPersistedGrantStore
+    {
+        IPersistGrantRepository _persistGrantRepository;
+        IUnitOfWork _unitOfWork;
+
+        public CustomPersistedGrantStore(IPersistGrantRepository persistGrantRepository, IUnitOfWork unitOfWork)
+        {
+            _persistGrantRepository = persistGrantRepository;
+            _unitOfWork = unitOfWork;
+        }
+
+        public async Task<IEnumerable<PersistedGrant>> GetAllAsync(PersistedGrantFilter filter)
+        {
+            filter.Validate();
+
+            var persistedGrants = await Filter(_persistGrantRepository.GetAll(), filter).ToArrayAsync();
+            persistedGrants = Filter(persistedGrants.AsQueryable(), filter).ToArray();
+
+            var model = persistedGrants.Select(x => x.ToModel());
+
+            //Logger.LogDebug("{persistedGrantCount} persisted grants found for {@filter}", persistedGrants.Length, filter);
+
+            return model;
+        }
+
+        public async Task<PersistedGrant> GetAsync(string key)
+        {
+            var persistedGrant = (await _persistGrantRepository.GetAll().Where(x => x.Key == key).ToArrayAsync())
+               .SingleOrDefault(x => x.Key == key);
+            var model = persistedGrant?.ToModel();
+
+            //Logger.LogDebug("{persistedGrantKey} found in database: {persistedGrantKeyFound}", key, model != null);
+
+            return model;
+        }
+
+        public async Task RemoveAllAsync(PersistedGrantFilter filter)
+        {
+            filter.Validate();
+
+            var persistedGrants = await Filter(_persistGrantRepository.GetAll(), filter).ToArrayAsync();
+            persistedGrants = Filter(persistedGrants.AsQueryable(), filter).ToArray();
+
+            //Logger.LogDebug("removing {persistedGrantCount} persisted grants from database for {@filter}", persistedGrants.Length, filter);
+
+            
+
+            foreach (var item in persistedGrants)
+            {
+                _persistGrantRepository.Remove(item);
+            }
+
+            try
+            {
+                await _unitOfWork.SaveChangesAsync();
+            }
+            catch (DbUpdateConcurrencyException ex)
```

Comment:
Why this exception will be thrown? Do we have optimistic concurrency token in DB Entity? Add token, or remove try/catch